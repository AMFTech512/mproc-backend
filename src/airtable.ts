import Airtable from "airtable";
import Joi from "joi";

export interface AirtableConfig {
  apiKey: string;
}

export function getAirtableConfig(): AirtableConfig {
  return Joi.attempt(
    {
      apiKey: process.env.AIRTABLE_API_KEY,
    },
    Joi.object({
      apiKey: Joi.string().required(),
    })
  );
}

export function initAirtableClient(config?: AirtableConfig) {
  const _config = config || getAirtableConfig();
  return new Airtable({ apiKey: _config.apiKey });
}
