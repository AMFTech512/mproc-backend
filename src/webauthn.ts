import { generateRegistrationOptions } from "@simplewebauthn/server";
import Joi from "joi";
import { AuthenticatorRow } from "./webauthn-authenticator-repo";

export interface WebAuthnConfig {
  rpName: string;
  rpID: string;
  attestationType: AttestationConveyancePreference;
}

export function getWebAuthnConfig(): WebAuthnConfig {
  return Joi.attempt(
    {
      rpName: process.env.WEBAUTHN_RP_NAME,
      rpID: process.env.WEBAUTHN_RP_ID,
      attestationType: process.env.WEBAUTHN_ATTESTATION_TYPE,
    },
    Joi.object({
      rpName: Joi.string().required(),
      rpID: Joi.string().required(),
      attestationType: Joi.string()
        .valid("direct", "enterprise", "indirect", "none")
        .default("none"),
    })
  );
}

export async function getRegistrationOptions(
  config: WebAuthnConfig,
  userID: string,
  userName: string,
  existingAuthenticators?: AuthenticatorRow[]
) {
  return await generateRegistrationOptions({
    rpName: config.rpName,
    rpID: config.rpID,
    userID,
    userName,
    // Don't prompt users for additional information about the authenticator
    // (Recommended for smoother UX)
    attestationType: config.attestationType,
    excludeCredentials: existingAuthenticators?.map((authenticator) => ({
      id: authenticator.credentialID,
      type: "public-key",
      // Optional
      transports: authenticator.transports,
    })),
  });
}
