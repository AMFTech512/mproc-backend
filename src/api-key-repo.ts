import { Client } from "pg";
import { DIContainer } from "./di";
import { generateSecureRandomString } from "./util";
import { createHash } from "crypto";

export interface ApiKeyRow {
  key_hash: Buffer;
  owner_id: string;
  created_at: Date;
  name: string;
  is_active: boolean;
}

export class ApiKeyRepo {
  private _dbClient: Client;

  constructor(container: DIContainer) {
    this._dbClient = container.postgresClient;
  }

  /**
   * Generates a random key and its hash.
   * @returns A tuple of [key, key hash].
   */
  static generateKeyHashPair(): [string, Buffer] {
    const key = generateSecureRandomString(16);
    const keyHash = ApiKeyRepo.hashKey(key);

    return [key, keyHash];
  }

  static hashKey(key: string): Buffer {
    return createHash("sha256").update(key).digest();
  }

  async insert(row: ApiKeyRow) {
    return await this._dbClient.query(
      "INSERT INTO api_keys (key_hash, owner_id, created_at, name, is_active) VALUES ($1, $2, $3, $4, $5)",
      [row.key_hash, row.owner_id, row.created_at, row.name, row.is_active]
    );
  }

  async getByKeyHash(keyHash: Buffer) {
    const res = await this._dbClient.query(
      "SELECT * FROM api_keys WHERE key_hash = $1",
      [keyHash]
    );
    return res.rows[0] as ApiKeyRow;
  }

  async deleteByKeyHash(keyHash: Buffer) {
    return await this._dbClient.query(
      "DELETE FROM api_keys WHERE key_hash = $1",
      [keyHash]
    );
  }
}
