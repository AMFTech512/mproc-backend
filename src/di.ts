import { Application } from "express";
import PQueue from "p-queue";
import { Client } from "pg";
import { JwtConfig } from "./jwt";
import { UserRepo } from "./user-repo";
import { ApiKeyRepo } from "./api-key-repo";
import { KeyUsageLogRepo } from "./key-usage-repo";
import { WebAuthnConfig } from "./webauthn";
import { WebAuthnChallengeRepo } from "./webauthn-challenge-repo";
import { WebAuthnAuthenticatorRepo } from "./webauthn-authenticator-repo";

export interface DIContainer {
  expressApp: Application;
  postgresClient: Client;
  userRepo: UserRepo;
  apiKeyRepo: ApiKeyRepo;
  keyUsageLogRepo: KeyUsageLogRepo;
  webAuthnChallengeRepo: WebAuthnChallengeRepo;
  webAuthnAuthenticatorRepo: WebAuthnAuthenticatorRepo;
  imagePQueue: PQueue;
  jwtConfig: JwtConfig;
  webAuthnConfig: WebAuthnConfig;
}

export function createDIContainer(): DIContainer {
  return {} as DIContainer;
}
