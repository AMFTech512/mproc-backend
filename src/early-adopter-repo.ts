import { Base } from "airtable";
import { DIContainer } from "./di";

export class EarlyAdopterRepo {
  private _base: Base;

  constructor(container: DIContainer) {
    // TODO: move this into a config?
    this._base = container.airtableClient.base("appwx5TpNTb8PdfYi");
  }

  async insertEarlyAdopter(email: string) {
    // TODO: also move this table name into a config?
    return await this._base("Early adopters").create([{ fields: { email } }]);
  }
}
