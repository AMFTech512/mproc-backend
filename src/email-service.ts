import { Resend } from "resend";
import { DIContainer } from "./di";

const DEFAULT_FROM_EMAIL = "Austin Fay <austin@mproc.io>";
const DEFAULT_AUDIENCE_ID = "cf9fb491-4cf4-48f1-98fa-e3c6a6d847d3";

export interface NewContactOpts {
  email: string;
  firstName?: string;
  lastName?: string;
  audienceId?: string;
  unsubscribed?: false;
}

export class EmailService {
  private _resendClient: Resend;

  constructor(container: DIContainer) {
    this._resendClient = container.resendClient;
  }

  async addContact(opts: NewContactOpts) {
    const res = await this._resendClient.contacts.create({
      email: opts.email,
      firstName: opts.firstName,
      lastName: opts.lastName,
      audienceId: opts.audienceId ?? DEFAULT_AUDIENCE_ID,
      unsubscribed: opts.unsubscribed ?? false,
    });

    if (res.error) {
      throw res.error;
    }
  }

  async sendEmail(
    toEmail: string,
    subject: string,
    html: string,
    fromEmail = DEFAULT_FROM_EMAIL
  ) {
    const res = await this._resendClient.emails.send({
      to: toEmail,
      from: fromEmail,
      subject,
      html,
    });

    if (res.error) {
      throw res.error;
    }
  }
}
