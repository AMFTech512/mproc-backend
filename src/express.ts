import express from "express";
import { DIContainer } from "./di";
import Joi from "joi";
import multer from "multer";
import cors from "cors";
import { handleUpload, handleUserCreate } from "./handlers";
import packageJson from "../package.json";

interface ServerConfig {
  port: number;
}

function getServerConfig(): ServerConfig {
  return Joi.attempt(
    {
      port: process.env.SERVER_PORT,
    },
    Joi.object({
      port: Joi.number().required(),
    })
  );
}

export function initExpressApp(container: DIContainer) {
  const app = express();
  const config = getServerConfig();

  const upload = multer({
    dest: "./uploads",
  });

  app.use(cors());

  app.get("/", (req, res) => {
    res.send(`v${packageJson.version} - OK`);
  });

  app.post("/upload", upload.single("file"), handleUpload(container));

  app.post("/user", express.json(), handleUserCreate(container));

  app.listen(config.port, () => {
    console.log(`Server is running on port ${config.port}`);
  });

  return app;
}
