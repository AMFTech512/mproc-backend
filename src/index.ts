import { initAirtableClient } from "./airtable";
import { ApiKeyRepo } from "./api-key-repo";
import { createDIContainer } from "./di";
import { EarlyAdopterRepo } from "./early-adopter-repo";
import { initExpressApp } from "./express";
import { getJwtConfig } from "./jwt";
import { KeyUsageLogRepo } from "./key-usage-repo";
import { initQueue } from "./p-queue";
import { initPostgresClient } from "./postgres";
import { UserRepo } from "./user-repo";
import { getWebAuthnConfig } from "./webauthn";
import { WebAuthnAuthenticatorRepo } from "./webauthn-authenticator-repo";
import { WebAuthnChallengeRepo } from "./webauthn-challenge-repo";

await bootstrap();

async function bootstrap() {
  console.log("Bootstrapping application...");

  // create the DI container
  console.log("Creating DI container...");
  const container = createDIContainer();

  // initialize the jwt config
  console.log("Initializing jwt config...");
  container.jwtConfig = getJwtConfig();

  // initialize the webauthn config
  console.log("Initializing webauthn config...");
  container.webAuthnConfig = getWebAuthnConfig();

  // initialize the postgres client
  console.log("Initializing postgres client...");
  container.postgresClient = await initPostgresClient();

  // initialize the airtable client
  console.log("Initializing airtable client...");
  container.airtableClient = initAirtableClient();

  // initialize the user repo
  console.log("Initializing user repo...");
  container.userRepo = new UserRepo(container);

  // initialize api key repo
  console.log("Initializing api key repo...");
  container.apiKeyRepo = new ApiKeyRepo(container);

  // initialize the key usage log repo
  console.log("Initializing key usage log repo...");
  container.keyUsageLogRepo = new KeyUsageLogRepo(container);

  // initialize webauthn challenge repo
  console.log("Initializing webauthn challenge repo...");
  container.webAuthnChallengeRepo = new WebAuthnChallengeRepo(container);

  // initialize the webauthn authenticator repo
  console.log("Initializing webauthn authenticator repo...");
  container.webAuthnAuthenticatorRepo = new WebAuthnAuthenticatorRepo(
    container
  );

  // initialize the early adopter repo
  console.log("Initializing early adopter repo...");
  container.earlyAdopterRepo = new EarlyAdopterRepo(container);

  // initialize the image process queue
  console.log("Initializing process queue...");
  container.imagePQueue = initQueue();

  // initialize the express app
  console.log("Initializing express app...");
  container.expressApp = initExpressApp(container);
}
