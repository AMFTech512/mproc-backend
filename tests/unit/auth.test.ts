import { describe, expect, it } from "bun:test";
import { getRegistrationOptions } from "../../src/webauthn";

describe("Auth", () => {
  describe("getRegistrationOptions", () => {
    it("should return registration options", async () => {
      const options = await getRegistrationOptions(
        {
          attestationType: "none",
          rpID: "localhost",
          rpName: "localhost",
        },
        "123",
        "bob"
      );

      expect(options).toMatchSnapshot();
    });
  });
});
