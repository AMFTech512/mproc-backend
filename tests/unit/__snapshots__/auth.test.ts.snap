// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth getRegistrationOptions should return registration options 1`] = `
{
  "attestation": "none",
  "authenticatorSelection": {
    "requireResidentKey": false,
    "residentKey": "preferred",
    "userVerification": "preferred",
  },
  "challenge": "HzV6dk2zkmcZZ9bwSgQXrN2leMGLzFWKmCjx-bnE6vc",
  "excludeCredentials": [],
  "extensions": {
    "credProps": true,
  },
  "pubKeyCredParams": [
    {
      "alg": -8,
      "type": "public-key",
    },
    {
      "alg": -7,
      "type": "public-key",
    },
    {
      "alg": -257,
      "type": "public-key",
    },
  ],
  "rp": {
    "id": "localhost",
    "name": "localhost",
  },
  "timeout": 60000,
  "user": {
    "displayName": "bob",
    "id": "123",
    "name": "bob",
  },
}
`;
